library/alloc/src/vec/mod.rs

```diff
@@ -2792,6 +2792,51 @@ impl<T, I: SliceIndex<[T]>, A: Allocator> IndexMut<I> for Vec<T, A> {
     }
 }
 
+/// Collects an iterator into a Vec, commonly via [`Iterator::collect()`]
+///
+/// # Allocation behavior
+///
+/// In general `Vec` does not guarantee any particular grow/allocation stategy.
+/// That also applies to this trait impl.
+///
+/// **Note:** This section covers implementation details and is therefore exempt from
+/// stability guarantees.
+///
+/// Vec may use any or none of the following strategies,
+/// depending on the supplied iterator:
+///
+/// * preallocate based on [`Iterator::size_hint()`]
+///   * and panic if the number of items is not outside the provided lower/upper bounds
```

Comment:
I assume you mean if it *is* outside -- but what's an example where the current implementation would panic?

Reply:
This is an example what it might do, not what it currently does, in the "these are not precise guarantees" spirit.

Reply:
OK, well that kind of introduces a bogeyman, and I'm not sure we would ever want to do that. Did you have a *possible* scenario in mind?

Reply:
When the size hint bounds are tight we can treat them as exact length, only allocate once and bail if the iterator lied to us, which means it's only a single call to the allocator and single-pass loop.

Or something similar to the `TrustedLen` collect impl that uses the lower bound to run part of the collection as a counted loop and ignoring the discriminant of the `Option`. With enum `offset_of` we can fish the bytes out of the enum even if it is `None`, copy that into the `Vec` and only check later if we actually got the amount of expected items. Then if things don't match up we'll have to panic because there's garbage in the Vec.

Whether any of those things are actually beneficial will have to be measured.

Reply:
```suggestion
///   * and panic if the number of items is outside the provided lower/upper bounds
```

Reply:
Ah sorry, missed that the first time.